### PR TITLE
[WIP] Fb link event element

### PR DESCRIPTION
**Pull request by LeonardBesseau**
Create event when creating element.

Need to add link when changing element and soft deleting element

### CODE REVIEW

File: controllers/elementController.js [hunk 1]
```diff
@@ -1,13 +1,167 @@
...
+
+            await ElementEvent.create({
+                user: req.user.id,
+                element: doc._id,
+                change: "Move",
+                oldLocation: doc.idLocation
+
+            });
+        }).then(dat => {
+            doc.idLocation = req.params.location;
```
Comment (K-do) [inline controllers/elementController.js:140]: ?

File: models/eventModel.js [hunk 2]
```diff
@@ -67,12 +67,12 @@ const eventElement = new mongoose.Schema({
...
-            validator: function () {
+        required: [
+            function () {
                 return this.change === "Move";
             },
-            message: "Location is required"
-        },
+            "Location is required"
+        ],
         validate: {
```
Comment (K-do) [inline models/eventModel.js:76]: Difference between required and validate ?

File: controllers/elementController.js [hunk 3]
```diff
@@ -1,13 +1,167 @@
...
+    );
+    if(!location){
+        return next(new AppError(404, 'fail', 'IdLocation not found'), req, res, next);
+    }
+    try {
+        const session = await mongoose.startSession();
+        let doc = {}
+        await session.withTransaction(async () => {
+            doc = await Element.findByIdAndUpdate(req.params.id, {
+                idLocation: req.body.idLocation
```
Comment (K-do) [inline controllers/elementController.js:126]: should be req.params.location, or location._id ?

File: tests/controller/elementControllerTest.js [hunk 4]
```diff
@@ -125,15 +131,21 @@ before(async function () {
...
     }).then((doc) => {
         return doc._id;
     });
     idQR2 = await QR.create({
-        code: "QRcode2"
+        code: codeQR2
     }).then((doc) => {
         return doc._id;
     });
+    await QR.create({
```
Comment (K-do) [inline tests/controller/elementControllerTest.js:143]: idQR3 never assigned

File: tests/controller/elementControllerTest.js [hunk 5]
```diff
@@ -125,15 +131,21 @@ before(async function () {
...
+        code: codeQR3
+    });
+    idQR4 =await QR.create({
```
Comment (K-do) [inline tests/controller/elementControllerTest.js:146]: Forgot .then((doc) => {return doc_id;} );

File: tests/controller/elementControllerTest.js [hunk 6]
```diff
@@ -163,6 +175,14 @@ before(async function () {
     }).then((doc) => {
         return doc._id;
     });
+    await Element.create({
```
Comment (K-do) [inline tests/controller/elementControllerTest.js:178]: Where is idElement1 ?

File: tests/controller/elementControllerTest.js [hunk 7]
```diff
@@ -154,7 +166,7 @@ before(async function () {
         return doc._id;
     });
     idElement2 = await Element.create({
-        idQR: idQR2,
```
Comment (K-do) [inline tests/controller/elementControllerTest.js:157]: Is this idElement1 ?

File: tests/controller/elementControllerTest.js [hunk 8]
```diff
@@ -339,7 +437,7 @@ describe('elementController', function () {
                     });
                     expect(res.status).to.be.equal(200);
                     expect(res.body.status).to.be.equal("success");
-                    expect(res.body.results).to.be.equal(2);
+                    expect(res.body.results).to.be.equal(3);
```
Comment (K-do) [inline tests/controller/elementControllerTest.js:440]: Why is it 3 ?

**Review by K-do — CHANGES_REQUESTED** [review]
Some parts are not clear

File: tests/controller/elementControllerTest.js (hunk 4 above)
Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:143]: No needed. Just need to add a QR in the database

File: tests/controller/elementControllerTest.js (hunk 5 above)
Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:146]: Yes. Good job for seeing that one.

Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:146]: Corrected in next commit

File: tests/controller/elementControllerTest.js (hunk 6 above)
Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:178]: No need

File: tests/controller/elementControllerTest.js (hunk 8 above)
Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:440]: The number of element that was present during my test. I checked and after each only clean element with price of zero.
image
Get All takes element that are non active by default. Should we change that ?

File: tests/controller/elementControllerTest.js (hunk 7 above)
Comment (LeonardBesseau) [reply tests/controller/elementControllerTest.js:157]: Not sure to understand

Comment (K-do) [reply tests/controller/elementControllerTest.js:157]: you never asigned idElement1 but it is used later for soft/hard dlete tests

File: tests/controller/elementControllerTest.js (hunk 8 above)
Comment (K-do) [reply tests/controller/elementControllerTest.js:442]: hmm to discuss, if we change, we'll need a second methode to treat non-active

File: models/eventModel.js (hunk 2 above)
Comment (LeonardBesseau) [reply models/eventModel.js:76]: Require built-in validator for checking presence if true,
Validate custom validator to verify properties of the data

File: controllers/elementController.js (hunk 3 above)
Comment (LeonardBesseau) [reply controllers/elementController.js:126]: req.params.location

**Review by K-do — APPROVED** [review]

**Pull request merged**